storage/provider/managedfs.go

```diff
@@ -272,31 +273,96 @@ func extractMtabEntry(etcDir string, devicePath, mountPoint string) (string, err
 
 // addFstabEntry creates an entry in /etc/fstab for the specified
 // device path and mount point so long as there's no existing entry already.
-func addFstabEntry(etcDir string, devicePath, mountPoint, entry string) error {
-	f, err := os.OpenFile(filepath.Join(etcDir, "fstab"), os.O_CREATE|os.O_APPEND|os.O_RDWR, 0644)
-	if err != nil {
+func addFstabEntry(etcDir, devicePath, UUID, mountPoint, entry string) error {
+	f, err := os.Open(filepath.Join(etcDir, "fstab"))
+	if err != nil && !os.IsNotExist(err) {
 		return errors.Annotate(err, "opening /etc/fstab")
 	}
-	defer f.Close()
+	if err == nil {
+		defer f.Close()
+	}
 
-	// Ensure there's no entry there already
+	newFsTab, err := ioutil.TempFile(etcDir, "juju-fstab-")
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer func() {
+		newFsTab.Close()
+		os.Remove(newFsTab.Name())
+	}()
+	if err := os.Chmod(newFsTab.Name(), 0644); err != nil {
+		return errors.Trace(err)
+	}
+
+	// Add nofail if not there already
+	resultFields := strings.Fields(entry)
+	options := set.NewStrings()
+	if len(resultFields) >= 4 {
+		options = set.NewStrings(strings.Split(resultFields[3], ",")...)
+	}
+	if !options.Contains("nofail") {
+		options.Add("nofail")
+		opts := strings.Join(options.SortedValues(), ",")
+		if len(resultFields) >= 4 {
+			resultFields[3] = opts
+		} else {
+			resultFields = append(resultFields, opts)
+		}
+	}
+
+	uuidField := "UUID=" + UUID
+	found := false
+	// Scan all the fstab lines, searching for one
+	// which describes the entry we want to create.
 	scanner := bufio.NewScanner(f)
-	for scanner.Scan() {
+	for f != nil && scanner.Scan() {
 		line := scanner.Text()
 		fields := strings.Fields(line)
-		if len(fields) >= 2 && fields[0] == devicePath && fields[1] == mountPoint {
-			return nil
+		if len(fields) < 2 || fields[1] != mountPoint {
+			goto writeLine
+		}
+		// Is the line the UUID based mount entry we want.
+		if fields[0] == uuidField {
+			found = true
+			goto writeLine
+		}
+		// Is the line for some other entry.
+		if fields[0] != devicePath {
+			goto writeLine
+		}
+		// We have a match, if UUID is not yet known, retain the line.
+		if UUID == "" {
+			found = true
+			goto writeLine
+		}
+		continue
+	writeLine:
+		_, err := newFsTab.WriteString(line + "\n")
+		if err != nil {
+			return errors.Trace(err)
 		}
 	}
 	if err := scanner.Err(); err != nil {
 		return errors.Trace(err)
 	}
 
-	// The entry will be written at the end of the fstab file.
-	if _, err = f.WriteString("\n" + entry + "\n"); err != nil {
-		return errors.Annotate(err, "writing /etc/fstab")
+	if !found {
```

Comment:
This is not a real notfound, it could be a dev path entry found or not found.
This `found` var is a bit confusing to read

Reply:
renamed found to addNewEntry